basiclti/web-ifp/src/java/org/sakaiproject/portlets/SakaiIFrame.java

```diff
@@ -233,9 +243,68 @@ public void doView(RenderRequest request, RenderResponse response)
 			} else {
 				out.println(rb.getString("get.info.notconfig"));
 			}
+		}
+
 
-			// System.out.println("==== doView complete ====");
+	/**
+	 * Patch the content item if it was copied from the !user template.
+	 *
+	 * We only do this once is there is a source, and we cannot get it,
+	 * we either make a new content item from the tool or we empty the
+	 * source property.
+	 */
+	public Map<String, Object> patchContentItem(Long key, Placement placement)
+	{
+		// Get out tool configuration so we can fix things up...
+		ToolConfiguration toolConfig = SiteService.findTool(placement.getId());
+
+		// Look up the content item, bypassing authz checks
+		Map<String, Object> content = m_ltiService.getContentDao(key);
+		if ( content == null ) return null;
+		Long tool_id = getLongNull(content.get("tool_id"));
+
+		// Look up the tool associated with the Content Item
+		// checking Authz to see is we can touch this tool
+		Map<String, Object> tool = m_ltiService.getTool(tool_id, placement.getContext());
+		if ( tool == null ) return null;
+
+		// Now make a content item from this tool inheriting from the other content item
+		Properties props = new Properties();
+		for (Map.Entry<String, Object> entry : content.entrySet()) {
+			String k = entry.getKey();
+			Object value = entry.getValue();
+			if ( value == null ) continue;
+			if ( k.endsWith("_at") ) continue;
+			props.put(k, value.toString());
 		}
+		props.put(LTIService.LTI_TOOL_ID, tool_id.toString());
+		props.put(LTIService.LTI_SITE_ID, placement.getContext());
+		props.put(LTIService.LTI_PLACEMENT, placement.getId());
+
+		Object retval = m_ltiService.insertContent(props, placement.getContext());
+		if ( retval instanceof String ) {
+			M_log.error("Unable to insert LTILinkItem tool={} placement={}",tool_id,placement.getId());
+			placement.getPlacementConfig().setProperty(SOURCE,"");
+			placement.save();
+			return null;
```

Comment:
Typically if your method is to return a Map then its good form to always return a Map even if its an empty Map, see java.util.Collections.emptyMap().

If this were a private method it wouldn't be as important but since its public!